### PR TITLE
tests: net: lib: coap: Added new functions to improve code coverage

### PR DESCRIPTION
**Pull request by tarunkum**
In coap module previous test cases are modified as well as
new test functions are added to increase the code coverage

Signed-off-by: Tarun Kumar Agarwal <tarun.kumar.agarwal@intel.com>

### CODE REVIEW

**Review by mike-scott — COMMENTED** [review]
Commit description is fairly lacking for an addition of ~330 lines of code.  Also, I would break this up into several commits for easier review.

File: tests/net/lib/coap/src/main.c [hunk 1]
```diff
@@ -66,6 +74,47 @@ static struct sockaddr_in6 dummy_addr = {
...
+	struct net_buf *frag;
+	int r;
+
+	static struct net_context *context;
+
+	pkt = net_pkt_get_tx(context, K_FOREVER);
+	frag = net_pkt_get_data(context, K_FOREVER);
+	net_pkt_frag_add(pkt, frag);
+
+	r = coap_well_known_core_get(resource, request, &response, pkt);
```
Comment (rveerama1) [inline tests/net/lib/coap/src/main.c]: Do not only concentrate on code coverage, these are unit tests. Also test some parsed information like parsed and expected information is matching or not.

File: tests/net/lib/coap/src/main.c [hunk 2]
```diff
@@ -66,6 +74,47 @@ static struct sockaddr_in6 dummy_addr = {
...
+	if (r < 0) {
+		net_pkt_unref(response.pkt);
+		return r;
+	}
+	get_from_ip_addr(request, &from);
```
Comment (rveerama1) [inline tests/net/lib/coap/src/main.c]: This is not required IMO. You retrieved 'from' address, then?

File: tests/net/lib/coap/src/main.c [hunk 3]
```diff
@@ -116,6 +165,8 @@ static int test_build_empty_pdu(void)
 	return result;
 }
 
+/* FIXME */
+#define BLOCK_WISE_TRANSFER_SIZE_GET 2048
```
Comment (rveerama1) [inline tests/net/lib/coap/src/main.c]: IMO, its not belongs here.

File: tests/net/lib/coap/src/main.c [hunk 4]
```diff
@@ -235,6 +288,8 @@ static int test_parse_empty_pdu(void)
 		goto done;
 	}
 
+	get_from_ip_addr(&cpkt, &from);
```
Comment (rveerama1) [inline tests/net/lib/coap/src/main.c]: Not required.

File: tests/net/lib/coap/src/main.c [hunk 5]
```diff
@@ -260,6 +315,15 @@ static int test_parse_empty_pdu(void)
 		goto done;
 	}
 
+	if (coap_header_get_type(&cpkt) == COAP_TYPE_RESET) {
```
Comment (rveerama1) [inline tests/net/lib/coap/src/main.c]: This test case is to test simple PDU packet, not related to Observer functionality. Do not mix here. Write a separate test case to verify Observer functionality.

File: tests/net/lib/coap/src/main.c [hunk 6]
```diff
@@ -350,6 +417,11 @@ static int test_parse_empty_pdu_1(void)
 		goto done;
 	}
 
+	r = coap_find_options(&cpkt, COAP_OPTION_OBSERVE, option, count);
```
Comment (rveerama1) [inline tests/net/lib/coap/src/main.c]: Ditto comments related to Observer functionality.

File: tests/net/lib/coap/src/main.c [hunk 7]
```diff
@@ -855,6 +926,28 @@ static int test_observer_server(void)
 		goto done;
 	}
 
+	u16_t offset;
+	u16_t len;
```
Comment (rveerama1) [inline tests/net/lib/coap/src/main.c]: Declare variables in the beginning.

File: tests/net/lib/coap/src/main.c [hunk 8]
```diff
@@ -1551,11 +1826,53 @@ static int test_parse_malformed_marker(void)
 	return result;
 }
 
+static int test_token(void)
```
Comment (rveerama1) [inline tests/net/lib/coap/src/main.c]: Need not to write separate test case for only token. Version, type, code, id, token and etc.. will be there in every coap packet. Verify all those information in one of the test case. That will be enough I guess.

**Review by rveerama1 — CHANGES_REQUESTED** [review]
This PR mixed up things. Trying to increase the code coverage but missing the core use of unit tests. Also mixed up tests (like checking observe functionality in non related test cases) and some coding guidelines.

File: tests/net/lib/coap/src/main.c (hunk 8 above)
Comment (tarunkum) [reply tests/net/lib/coap/src/main.c]: Taken care

**Comment by tarunkum** [comment]
@rveerama1 Please review this code. From now onwards I will pushing the code in chunks & will work on your provided comments. Thanks a lot for your valuable comments.

File: tests/net/lib/coap/src/main.c [hunk 9]
```diff
@@ -1197,6 +1197,187 @@ static int test_block_size(void)
...
+		goto done;
+	}
+
+	req.offset = sizeof(ipv6_block);
+
+	coap_block_transfer_init(&req_ctx, COAP_BLOCK_64, 255);
+
+	r = coap_append_block2_option(&req, &req_ctx);
+	if (r) {
+		TC_PRINT("Unable to append block1 option\n");
```
Comment (rveerama1) [inline tests/net/lib/coap/src/main.c]: Wrong debug print; you are trying to add Block2 option.

File: tests/net/lib/coap/src/main.c [hunk 10]
```diff
@@ -1197,6 +1197,187 @@ static int test_block_size(void)
...
+		goto done;
+	}
+
+	r = coap_append_size2_option(&req, &req_ctx);
+	if (r) {
+		TC_PRINT("Unable to append size1 option\n");
```
Comment (rveerama1) [inline tests/net/lib/coap/src/main.c]: Wrong debug print; you are trying to add Size2 option.

File: tests/net/lib/coap/src/main.c [hunk 11]
```diff
@@ -1197,6 +1197,187 @@ static int test_block_size(void)
...
+	r = coap_packet_append_payload_marker(&req);
+	if (r) {
+		TC_PRINT("Unable to append payload marker\n");
+		goto done;
+	}
+
+	r = coap_packet_append_payload(&req, payload,
+				       coap_block_size_to_bytes(COAP_BLOCK_64));
+	if (r) {
+		TC_PRINT("Unable to append payload marker\n");
```
Comment (rveerama1) [inline tests/net/lib/coap/src/main.c]: Wrong debug print; you are trying to add payload.

File: tests/net/lib/coap/src/main.c [hunk 12]
```diff
@@ -1197,6 +1197,187 @@ static int test_block_size(void)
...
+	if (r < 0) {
+		TC_PRINT("Unable to initialize request\n");
+		goto done;
+	}
+
+	req.offset = sizeof(ipv6_block);
+
+	r = coap_append_block2_option(&req, &req_ctx);
+	if (r) {
+		TC_PRINT("Unable to append block1 option\n");
```
Comment (rveerama1) [inline tests/net/lib/coap/src/main.c]: Ditto.

File: tests/net/lib/coap/src/main.c [hunk 13]
```diff
@@ -1197,6 +1197,187 @@ static int test_block_size(void)
...
+		goto done;
+	}
+
```
Comment (rveerama1) [inline tests/net/lib/coap/src/main.c]: Did you miss to add Size2 related option here?

File: tests/net/lib/coap/src/main.c [hunk 14]
```diff
@@ -1197,6 +1197,187 @@ static int test_block_size(void)
...
+	r = coap_packet_append_payload_marker(&req);
+	if (r) {
+		TC_PRINT("Unable to append payload marker\n");
+		goto done;
+	}
+
+	r = coap_packet_append_payload(&req, payload,
+				       coap_block_size_to_bytes(COAP_BLOCK_32));
+	if (r) {
+		TC_PRINT("Unable to append payload marker\n");
```
Comment (rveerama1) [inline tests/net/lib/coap/src/main.c]: Wrong debug print; you are trying to add payload.

**Review by rveerama1 — APPROVED** [review]
LGTM

**Pull request merged**